release-content/0.15/migration-guides/15320_Retained_Rendering.md

```diff
@@ -0,0 +1,96 @@
+
+### Extract components with `ExtractComponentPlugin`
+
+```
+// main world
+app.add_plugins(ExtractComponentPlugin::<ComponentToExtract>::default());
+```
+
+`ExtractComponentPlugin` has been changed to automatically sync entities with `ComponentToExtract`. This is done via the new `WorldSyncPlugin`. Any code using `ExtractComponentPlugin` will not require any changes.
+
+### Manual extraction using `Extract<Query<(Entity, ...)>>`
+
+```rust
+// in render world, inspired by bevy_pbr/src/cluster/mod.rs
+pub fn extract_clusters(
+    mut commands: Commands,
+    views: Extract<Query<(Entity, &Clusters, &Camera)>>,
+) {
+    for (entity, clusters, camera) in &views {
+        // some code
+        commands.get_or_spawn(entity).insert(...);
+    }
+}
+```
+An extract query in the render world queries for entities and components in the main world. Here `entity` is a main world entity and `get_or_spawn(main_world_entity).insert(...)` potentially inserts components on the wrong entity. Remember, there is no longer a one-to-one correspondence between the main and render world entities. Moreover `get_or_spawn` has been deprecated.
+
+In 0.15, you should use `RenderEntity` in place of `Entity` to get the correct entity in the render world. For entities to have a `RenderEntity` they need to be synced first. This can be done either via `WorldSyncPlugin` or adding the `SyncToRenderWorld` to the main world entity.
+
+This results in the following code:
+
+```rust
+// in render world, inspired by bevy_pbr/src/cluster/mod.rs
+pub fn extract_clusters(
+    mut commands: Commands,
+    views: Extract<Query<(RenderEntity, &Clusters, &Camera)>>,
+) {
+    for (render_entity, clusters, camera) in &views {
+        // some code
+        // After the sync step, all main world entities with a &RenderEntity have a corresponding (empty) render world entity. This should never panic.
+        commands.entity(render_entity).insert(...);
+    }
+}
+
+// in main world, when spawning
+world.spawn(Clusters::default(), Camera::default(), SyncToRenderWorld)
+```
+
+### Looking up main world entities in the render world
+
+In order to get the main world entity from a render world entity. It works much the same. Every synced render world entity has a `MainEntity` component you can query for that returns the correct main world entity.
+
+```rust
+// in the render world
+pub fn inspect_clusters(
+    views: Query<(MainEntity, &Clusters, &Camera)>
+) {
+    for (main_entity, clusters, camera in &views) {
+        // do something
+    }
+}
+```
+
+### General advice for working with main and render world entities
+
+When working with entities from both worlds it can be confusing. If you are every in a scenario where this isn't entirely clear (for example, when working on custom extraction code in the render world), we advice that you use `RenderEntity` and `MainEntity` as simple wrappers around `Entity`. Mixing these up can become a real headache and lead to some non-obvious errors.
```

Comment:
```suggestion
When working with entities from both worlds it can be confusing. If you are every in a scenario where this isn't entirely clear (for example, when working on custom extraction code in the render world), we advise that you use `RenderEntity` and `MainEntity` as simple wrappers around `Entity`. Mixing these up can become a real headache and lead to some non-obvious errors.
```